File: app/src/main/java/com/arcao/feedback/collector/Collector.java
package com.arcao.feedback.collector;

import java.io.PrintWriter;
import java.io.StringWriter;

public abstract class Collector {
	public abstract String getName();
	protected abstract String collect();

	@Override
	public String toString() {
		return "--- " + getName() + " ---\n" + collect() + "\n------\n\n";
	}

	protected String throwableToString(Throwable t) {
		StringWriter sw = new StringWriter();
		PrintWriter pw = new PrintWriter(sw);
		t.printStackTrace(pw);
		pw.close();
		return sw.toString();
	}
}

File: app/src/main/java/com/arcao/feedback/collector/Collector.kt
package com.arcao.feedback.collector

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.PrintWriter
import java.io.StringWriter

abstract class Collector {
    abstract val name: String

    abstract suspend fun collect(): String

    suspend operator fun invoke(): String = withContext(Dispatchers.IO) {
        "--- $name ---\n${collect()}\n------\n\n"
    }

    protected fun throwableToString(t: Throwable): String {
        val sw = StringWriter()
        PrintWriter(sw).use {
            t.printStackTrace(it)
        }
        return sw.toString()
    }
}
